src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -298,6 +298,52 @@ int sd_dhcp_client_set_client_id(sd_dhcp_client *client, uint8_t type,
         return 0;
 }
 
+int sd_dhcp_client_set_iaid_duid(sd_dhcp_client *client, uint32_t iaid,
+                                 uint16_t duid_type, uint8_t *duid, size_t duid_len) {
+        DHCP_CLIENT_DONT_DESTROY(client);
+        int r;
+        assert_return(client, -EINVAL);
+        zero(client->client_id);
+
+        client->client_id.type = 255;
+
+        /* If IAID is not configured, generate it. */
+        if (iaid == 0) {
+                r = dhcp_identifier_set_iaid(client->index, client->mac_addr,
+                                             client->mac_addr_len,
+                                             &client->client_id.ns.iaid);
+                if (r < 0)
+                        return r;
+        } else
+                client->client_id.ns.iaid = htobe32(iaid);
+
+        /* If DUID is not configured, generate DUID-EN. */
+        if (duid_len == 0) {
+                r = dhcp_identifier_set_duid_en(&client->client_id.ns.duid,
+                                                &duid_len);
+                if (r < 0)
+                        return r;
+        } else {
+                r = dhcp_validate_duid_len(client->client_id.type, duid_len);
+                if (r < 0)
+                        return r;
+                client->client_id.ns.duid.type = htobe16(duid_type);
+                memcpy(&client->client_id.ns.duid.raw.data, duid, duid_len);
+                duid_len += sizeof(client->client_id.ns.duid.type);
+        }
+
+        client->client_id_len = sizeof(client->client_id.type) + duid_len +
+                                sizeof(client->client_id.ns.iaid);
```

Comment:
No.  duid_len only holds the size of raw DUID data that gets copied to client_id.ns.duid.raw.data.

client_id_len should hold the size of the entire client identifier that goes on the wire: i.e sizeof(client identifier type) + sizeof(iaid) + duid size ( = sizeof(ns.duid.type) + duid_len ). 

Per RFC 4361: https://tools.ietf.org/html/rfc4361#section-6.1, this adds up to 19 bytes for DUID-EN. 

<img width="837" alt="screen shot 2016-03-25 at 1 31 27 pm" src="https://cloud.githubusercontent.com/assets/12750645/14053373/fd069a5e-f28d-11e5-9401-e8f8905f6070.png">